src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+        private static readonly int s_segmentShift = SegmentedArrayHelper.CalculateSegmentShift(s_segmentSize);
+        private static readonly int s_offsetMask = SegmentedArrayHelper.CalculateOffsetMask(s_segmentSize);
+
+        private readonly int _length;
+        private readonly T[][] _items;
+
+        public SegmentedArray(int length)
+        {
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            if (length == 0)
+            {
+                _items = Array.Empty<T[]>();
+                _length = 0;
+            }
+            else
+            {
+                _items = new T[(length + s_segmentSize - 1) >> s_segmentShift][];
+                for (var i = 0; i < _items.Length - 1; i++)
+                {
+                    _items[i] = new T[s_segmentSize];
+                }
+
+                _items[^1] = new T[length & s_offsetMask];
+                _length = length;
+            }
+        }
+
+        private SegmentedArray(int length, T[][] items)
+        {
+            _length = length;
+            _items = items;
+        }
+
+        public bool IsFixedSize => true;
+
+        public bool IsReadOnly => false;
+
+        public bool IsSynchronized => false;
+
+        public int Length => _length;
+
+        public object SyncRoot => _items ?? Array.Empty<T[]>().SyncRoot;
+
+        public ref T this[int index]
+        {
+            get
+            {
+                return ref _items[index >> s_segmentShift][index & s_offsetMask];
+            }
+        }
+
+        int ICollection.Count => _length;
+
+        int ICollection<T>.Count => _length;
+
+        int IReadOnlyCollection<T>.Count => _length;
+
+        T IReadOnlyList<T>.this[int index] => this[index];
+
+        T IList<T>.this[int index]
+        {
+            get => this[index];
+            set => this[index] = value;
+        }
+
+        object IList.this[int index]
+        {
+            get => this[index];
+            set => this[index] = (T)value;
+        }
+
+        public object Clone()
+        {
+            var items = (T[][])_items.Clone();
+            for (var i = 0; i < items.Length; i++)
+            {
+                items[i] = (T[])items[i].Clone();
+            }
+
+            return new SegmentedArray<T>(_length, items);
+        }
+
+        public void CopyTo(Array array, int index)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                _items[i].CopyTo(array, index + (i * s_segmentSize));
+            }
+        }
+
+        void ICollection<T>.CopyTo(T[] array, int arrayIndex)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                ICollection<T> collection = _items[i];
+                collection.CopyTo(array, arrayIndex + (i * s_segmentSize));
+            }
+        }
```

Comment:
i would think we'd need a CopyToNewLargerSegmentedArray helper.  i.e. if i have a list, and i want to grow it, i'll ned a new segmented list.  and i'll want to bulk copy the old values to the new.

Reply:
It will be implemented as a helper as part of `SegmentedList<T>`, along with mirrors of several other static members of `System.Array`.